src/librustc_typeck/collect.rs

```diff
@@ -1420,14 +1436,20 @@ fn explicit_predicates_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     // and the explicit where-clauses, but to get the full set of predicates
     // on a trait we need to add in the supertrait bounds and bounds found on
     // associated types.
-    if let Some((trait_ref, _)) = is_trait {
+    if let Some((trait_ref, ref items)) = is_trait {
         predicates = tcx.super_predicates_of(def_id).predicates;
 
-        // Add in a predicate that `Self:Trait` (where `Trait` is the
-        // current trait).  This is needed for builtin bounds.
-        predicates.push(trait_ref.to_poly_trait_ref().to_predicate());
+        debug!("explicit_predicates_of: 5 predicates={:?}", predicates);
+
+        if items.is_some() {
+            // Add in a predicate that `Self:Trait` (where `Trait` is the
+            // current trait).  This is needed for builtin bounds.
+            predicates.push(trait_ref.to_poly_trait_ref().to_predicate());
```

Comment:
I feel like we should be able to treat traits and trait aliases the same here if we set things up right (i.e., either both or neither would get this extra predicate). But it's not a big deal I don't think.

Reply:
Yeah, this is the part that works if you say "what are the preds of `TraitAlias`" (i.e. it adds only the super predicates and not `TraitAlias` itself) but falls over if you ask "what are the preds of <some item that refers to `TraitAlias`>", which may indicate it's too early or late to be distinguishing traits from aliases.